src/components/Popup/Popup.tsx

```diff
@@ -126,16 +127,45 @@ export default class Popup extends AutoControlledComponent<Extendable<PopupProps
 
   private static isBrowserContext = isBrowser()
 
+  private clickSubscription = EventStackSubscription.empty()
+  private triggerDomElement = null
+
   protected actionHandlers: AccessibilityActionHandlers = {
-    toggle: e => this.trySetOpen(!this.state.open, e, true),
-    closeAndFocusTrigger: e => {
-      this.trySetOpen(false, e, true)
-      _.invoke(this.state.target, 'focus')
+    toggle: e => {
+      this.trySetOpen(!this.state.open, e, true)
     },
+    closeAndFocusTrigger: e => this.closeAndFocusTrigger(e),
+  }
+
+  private closeAndFocusTrigger = e => {
+    if (this.state.open) {
+      this.trySetOpen(false, e, true)
+      _.invoke(this.triggerDomElement, 'focus')
+    }
+  }
+
+  private closeAndFocusTriggerOnClickIfOpen() {
+    if (this.state.open) {
+      setTimeout(() => this.clickSubscription.replaceWith('click', this.closeAndFocusTrigger))
```

Comment:
in your previous refactoring ```setTimeout``` was part of ```EventStack``` and, as I understood, a motivation of it was performance reason (please, correct me, if I am wrong). So the question is, should we leave it as part of the library so everyone who uses it can benefit from it? It can be as a separate method ```subscribeAsync```.

Reply:
this is a very good point. Actually, for now have decided to move this aspect to the client's code, as have discovered few additional subtle problems that may arise with eventStack's logic (and the fact that subscription handler could be called at the same JS frame). 

The simplest example - suppose that there is a following client code:

```js
foo() {
  const subscription = EventStack.subscribe('click', ...)
  subscription.unsubscribe()
}
```
Note that if subscription logic is async, this will result in subscription still being active at the end of the method - as 'unsubscribe' will be invoked at the moment when there is no subscription made yet.

One simplest solution would be to make all the necessary methods `async`, but this will complicate client's code. What I am thinking about, however, is a bit more elaborative approach where all these bits will be handled properly by eventStack, at the same time **without switch to async model** requested from the client (as async methods are quite error and race condition prone to be used in React event handlers). Now am working on the PR where would like to address this problem.